### PR TITLE
Implement multitarget compilation

### PR DESCRIPTION
**Pull request by steven-johnson**
This is the last piece needed for Issue #879 (runtime feature
detection). This needs some additional tests to be added, but the basic
infrastructure should all be in place — ready to start looking over
(but don’t merge yet).

### CODE REVIEW

**Comment by steven-johnson** [comment]
Updated tests, PTAL

File: src/Func.h [hunk 1]
```diff
@@ -509,6 +509,17 @@ class Func {
     EXPORT void compile_to_static_library(const std::string &filename_prefix, const std::vector<Argument> &args,
                                           const Target &target = get_target_from_environment());
 
+    /** Compile to static-library file and header pair once for each target;
+     * each resulting function will be evaluated (in order) via halide_can_use_target_features()
```
Comment (abadams) [inline src/Func.h]: I find this comment slightly confusing. Perhaps change "evaluated" to "considered"? Evaluated sounds like you're running the pipeline.

File: src/Generator.cpp [hunk 2]
```diff
@@ -251,22 +251,44 @@ int generate_filter_main(int argc, char **argv, std::ostream &cerr) {
...
-    const Target target = parse_target_string(target_string);
+    auto target_strings = split_string(target_string, ",");
+    std::vector<Target> targets;
+    for (const auto &s : target_strings) {
+        targets.push_back(parse_target_string(s));
+    }
 
     if (!runtime_name.empty()) {
+        if (targets.size() != 1) {
+            cerr << "Only one target allowed here";
```
Comment (abadams) [inline src/Generator.cpp]: either internal_error or user_error instead of cerr

File: src/Module.h [hunk 3]
```diff
@@ -112,6 +114,13 @@ EXPORT void compile_standalone_runtime(const std::string &object_filename, Targe
  */
 EXPORT Outputs compile_standalone_runtime(const Outputs &output_files, Target t);
 
+typedef std::function<Module(const std::string &, const Target &)> ModuleProducer;
+
+EXPORT void compile_multitarget(const std::string &fn_name, 
+                                const Outputs &output_files,
+                                const std::vector<Target> &targets, 
+                                ModuleProducer module_producer);
```
Comment (abadams) [inline src/Module.h]: Can this just take a vector of modules instead? That seems simpler than a vector of targets and a lambda.

File: src/Pipeline.h [hunk 4]
```diff
@@ -173,6 +173,17 @@ class Pipeline {
                                           const std::vector<Argument> &args,
                                           const Target &target = get_target_from_environment());
 
+    /** Compile to static-library file and header pair once for each target;
+     * each resulting function will be evaluated (in order) via halide_can_use_target_features()
```
Comment (abadams) [inline src/Pipeline.h]: If you change the comment in Func, change it here too

File: src/Generator.cpp (hunk 2 above)
Comment (steven-johnson) [inline src/Generator.cpp]: Everything else in this function uses cerr (it's essentially a main() function), so this is actually correct.

File: src/Func.h (hunk 1 above)
Comment (steven-johnson) [inline src/Func.h]: Done

File: src/Module.h (hunk 3 above)
Comment (steven-johnson) [inline src/Module.h]: Well, no: one caller needs to construct a new Generator for each target.

File: src/Pipeline.h (hunk 4 above)
Comment (steven-johnson) [inline src/Pipeline.h]: Done

**Comment by steven-johnson** [comment]
Buildbots are (finally) looking pretty good.

**Comment by abadams** [comment]
merging this and several other outstanding things locally, so I can test their combination for build breakages before pushing

**Pull request merged**